src/labthings_fastapi/thing_server_interface.py

```diff
@@ -242,16 +260,53 @@ def create_thing_without_server(
     if "thing_server_interface" in kwargs:
         msg = "You may not supply a keyword argument called 'thing_server_interface'."
         raise ValueError(msg)
-    return cls(
-        *args,
-        **kwargs,
-        thing_server_interface=MockThingServerInterface(
-            name=name, settings_folder=settings_folder
-        ),
-    )  # type: ignore[misc]
+
+    msi = MockThingServerInterface(name=name, settings_folder=settings_folder)
     # Note: we must ignore misc typing errors above because mypy flags an error
     # that `thing_server_interface` is multiply specified.
     # This is a conflict with *args, if we had only **kwargs it would not flag
     # any error.
     # Given that args and kwargs are dynamically typed anyway, this does not
     # lose us much.
+    thing = cls(*args, **kwargs, thing_server_interface=msi)  # type: ignore[misc]
+    if mock_all_slots:
+        _mock_slots(thing)
+    return thing
+
+
+def _mock_slots(thing: Thing) -> None:
+    """Mock the slots of a thing created by create_thing_without_server.
+
+    :param thing: The thing to mock the slots of
+    """
+    for attr_name, attr in class_attributes(thing):
+        if isinstance(attr, ThingSlot):
+            # Simply use the class of the first type that can be used.
+            mock_class = attr.thing_type[0]
+
+            # The names of the mocks we need to create to make a mapping of mock
+            # things for the slot to connect to.
+            mock_names = []
+            if attr.default is ...:
+                # if default use the name of the slot with mock
+                mock_names.append(f"mock-{attr_name}")
+            elif isinstance(attr.default, str):
+                mock_names.append(attr.default)
+            elif isinstance(attr.default, Iterable):
+                mock_names = list(attr.default)
+            # Note: If attr.default is None it will connect to None so no need for
+            # adding anything mapping of mocks.
+
+            # Populate a mapping of mocks pretending to be the things on the server
+            mocks = {}
+            for name in mock_names:
+                mock = Mock(spec=mock_class)
+                mock.name = name
+                mocks[name] = mock
+                # Store a copy of this mock in the mock server interface so it isn't
+                # garbage collected.
+                # Note that this causes mypy to throw an  `attr-defined` error as _mock
+                # only exists in the MockThingServerInterface
+                thing._thing_server_interface._mocks.append(mock)  # type: ignore[attr-defined]
+
+            attr.connect(thing, mocks, ...)
```

Comment:
Might it make sense to do this in a separate loop? Your code above does an impressively thorough job of constructing a collection of mock Things that you can pass to `attr.connect` without having to mess with the `ThingSlot` code (which I really like). However, the first slot will not see mocked Things that are added to satisfy later slots.

Splitting the loop means all the slots see all the mocked Things, which makes it function just a little bit more like the real thing, and might catch odd edge cases where there are interactions between multiple slots. I realise that is very edge-casey, but it's a pretty minor tweak so why not...

```suggestion
    for attr_name, attr in class_attributes(thing):
        if isinstance(attr, ThingSlot):
            attr.connect(thing, mocks, ...)
```